sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java

```diff
@@ -319,7 +324,13 @@ public void evaluate() throws Throwable {
         // the flow, preventing the enforcement(s) from being activated.
         // The motivation for this is avoiding enforcements over faulty pipelines.
         statement.evaluate();
-        enforcement.get().afterUserCodeFinished();
+
+        // If construction failed, then to reach here the test code must have caught the exception
+        // and succeeded. This suggests the test was verifying a specific construction error, and
+        // thus is expected.
+        if (!constructionHasFailed) {
```

Comment:
I understand now - because we mutate a `Pipeline` during application, we leave it in a corrupt state when expand of a composite fails (apparently primitives are OK...).

This means that `Pipeline` itself should have this bit and poison all future actions.